### ISSUE TITLE
MicroProfile Rest Client 2.0 (part of MP 4.0)

### ISSUE THREAD

**Issue by andymc12** [OpenLiberty/open-liberty#10786]
Describe the high level feature, including any external spec links.

When ready, add links to the Upcoming Feature Overview document and Feature Test Summary issue:
- UFO: https://ibm.box.com/s/u1b78kuj0v86vbpcuk47gfdlm4zhjqp1
- FTS: #14523 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval (Alasdair Nottingham) or N/A.
- [x] No Design / No UFO Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or **N/A**.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [ ] **FAT** - (Adam Yoho). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - (Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - (Steven Zvonek). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - (Don Bourne). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or **N/A**.
- [ ] **Blog** - (Laura Cowen) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by Emily-Jiang** [comment]
UFO review comments:

- Spec level

1. In MicroProfile Rest Client, add a new section on backward incompatible changes (see info [here](https://wiki.eclipse.org/MicroProfile/BreakingChanges)) under Release note for Rest Client 2.0.
1. Reactive Streams integration. Please investigate whether including RxJava is enough. If not, this needs to be optional if you have to force runtime to implement MP Reactive Streams Operators.
1. Redirect: 5 is the hop counter and the client needs to detect the loop.
1. SSE context propagation: need to specify what contexts are to be propagated. This is to be defined by context propagation or advise end users to specify what contexts they should propagate.

UFO comments:
1. The slide on End User Overview, need to remove the [] on param values
1. Open Liberty guides need to be updated to demonstrate the new feature
1. Jsonb instance needs to be static final for performance reasons.
1. Open API generator integration - needs to discuss with Martin Smithson off line to see whether there is integration concerns.
1. Should %2c can be used instead of comma? Andy is to investigate and update the UFO accordingly.

**Comment by andymc12** [comment]
All UFO comments have been addressed - updated version is now uploaded to box.
For issue 5, the answer is no - more info here (link included in UFO speaker notes): https://stackoverflow.com/a/34937992/6575578

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by andymc12** [comment]
Demo available at https://ibm.box.com/s/2jt8oxvdiywep6gdsmfb0gnp1izuzhb1

**Comment by CassTucker** [comment]
@andymc12  legal approves this version upgrade!

**Comment by ayoho** [comment]
@andymc12 It looks like https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.rest.client-2.0.feature#L9 needs to be updated to use the final version of the API rather than RC2. All other coordinates and usages look correct to me.

**Comment by andymc12** [comment]
@ayoho nice catch! I'm fixing this with PR #15449 - would you mind reviewing?  Thanks!

**Comment by donbourne** [comment]
@andymc12 , in your STE demo, did you demo the problem paths and see if there were any issues that could lead customers to have problems resolving those on their own?

**Comment by andymc12** [comment]
Opened doc issue https://github.com/OpenLiberty/docs/issues/3501 to make a minor update to [the REST Client page](https://openliberty.io/docs/20.0.0.12/rest-clients.html) to indicate that `Publisher` is also a valid asynchronous return type (for consuming SSEs).

**Comment by chirp1** [comment]
Andy and I agreed on the documentation update in ID issue #3501. Approving..

**Comment by hanczaryk** [comment]
Here are the SVT answers for https://github.com/OpenLiberty/open-liberty/issues/10786#issuecomment-744484734

a) Who conducted SVT tests for this feature?
Brian Hanczaryk

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
Yes, I agree that the serviceability of any problems encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without the need to engage L3.

**Comment by andymc12** [comment]
@donbourne 
> in your STE demo, did you demo the problem paths and see if there were any issues that could lead customers to have problems resolving those on their own?

Sadly, no - not during the STE (other than the trace strings to use, FFDCs to expect, etc.).  But I did demo some serviceability paths (using tcptunnel to view the HTTP requests/responses and using filters to gather additional diagnostics) in my demo to the EOI call (full demo link: https://ibm.box.com/s/44atzy5zyt97tvgv3u603nrw9p40610q - EOI call link: https://ibm.box.com/s/k819st21mqiibccv3pcsu71z5b9us0mb - both links may only be available to IBM employees).

**Comment by donbourne** [comment]
@andymc12 , using tcptunnel sounds like something beyond the normal capability we'd expect for a customer to diagnose problems they could run into.  Would that be needed for any of the common problem scenarios?  Is there anything that can be done to logging to avoid need for wire level debugging?

**Comment by andymc12** [comment]
>using tcptunnel sounds like something beyond the normal capability we'd expect for a customer to diagnose problems they could run into. Would that be needed for any of the common problem scenarios? Is there anything that can be done to logging to avoid need for wire level debugging?

@donbourne - users can get the same details by using trace, but IMO it's much easier to use something like tcptunnel (btw, tcptunnel is a lot easier to use than "packet sniffers" like wire shark or tcpdump, etc.).  For most of the common problem scenarios, existing logging should be sufficient.  You had asked about what was demo'd - from a serviceability perspective, I only demo'd tcptunnel.

**Comment by donbourne** [comment]
approving on basis rest client hasn't introduced new paths that would likely impact serviceability in 2.0 version.

**Comment by malincoln** [comment]
After discussion with team, resetting target to 21003 to working through following defects:
281216: Consistently failing on Java 8 builds. 
272083: Very consistent on all Windows platforms, including SOE 
276480: Intermittent, but still seems to pop up every few days, mostly on z/OS.
277806: Consistently fails on a couple Windows SOE definitions.

**Comment by andymc12** [comment]
In addition to the GA blog post (which should be based on the beta blog post), I have opened issue https://github.com/OpenLiberty/blogs/issues/1108 for a standalone blog post specifically for MP Rest Client 2.0.
